### PR TITLE
Improve int->long cast decomposition

### PR DESCRIPTION
**Pull request by mikedn**
If the cast is used by a multiply then zero extend instead of sign extend. This makes it easier for DecomposeMul to drop the high parts produced by the cast.

Contributes to #8839

### CODE REVIEW

**Comment by mikedn** [comment]
I'm a bit weary this approach because it relies on GT_MUL decomposition ignoring the high parts of the source operands. But it seems unlikely that GT_MUL decomposition will ever do things differently, the only reason a long GT_MUL reaches this far on 32 bit is exactly because it can be implemented with 32 bit instructions.

Another approach is to avoid introducing a new lclvar when the cast operand is already a lclvar. We're doing this in other places but it doesn't work too well, we get diffs like:
```asm
mov eax, [ebp-32]
mov edx, eax
```
becomes
```asm
mov eax, [ebp-32]
mov edx, [ebp-32]
```
And yet another approach would be, of course, to do proper dead code elimination somewhere after decomposition/lowering. But it seems preferable not to produce dead code in the first place. Besides, DCE won't remove the unnecessary lclvar.

This saves couple of hundred bytes in FX jit-diff, mostly from removing `mov ecx, 31` instructions that the removed cast leaves behind. Ironically, `StringBuider.Insert`, where this issue was found in the first place, increases by 5 bytes due to differences in register allocation.

cc @CarolEidt

File: src/jit/decomposelongs.cpp [hunk 1]
```diff
@@ -628,8 +628,20 @@ GenTree* DecomposeLongs::DecomposeCast(LIR::Use& use)
...
+            //
+            // If the cast is not used or it is used by a GT_MUL then we'll treat it as unsigned
+            // because doing so generates simpler IR. If a GT_MUL reaches here then it must be
+            // the GTF_MUL_64RSLT variant that uses only the low parts of its long operands, see
+            // DecomposeMul for details.
+            //
+
+            if (varTypeIsUnsigned(srcType) || use.IsDummyUse() || (use.User()->OperGet() == GT_MUL))
             {
+                if (!use.IsDummyUse() && (use.User()->OperGet() == GT_MUL))
```
Comment (pgavlin) [inline src/jit/decomposelongs.cpp]: Nit: you might enclose this entire `if` statement in `#ifdef DEBUG`.

**Comment by mikedn** [comment]
> but perhaps we should defer the work until DecomposeMul (e.g. pattern-match against the use of a cast by a multiply and early-out in DecomposeCast, then let DecomposeMul handle the pattern itself

Yep, I'll do that.

I think a cleaner solution would be to generate `GT_MUL_LONG` in morph and get rid of the casts and the weird `GTF_MUL_64RSLT` flag. But I don't feel brave enough to touch morph code :)

File: src/jit/decomposelongs.cpp (hunk 1 above)
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: Hmm, too many #ifdefs for too little benefit.

Comment (pgavlin) [inline src/jit/decomposelongs.cpp]: Fair enough. Hopefully the C++ compiler is smart enough to erase the if statement in a release build. :)

**Comment by danmosemsft** [comment]
Still working on this one @mikedn ?

**Comment by mikedn** [comment]
@danmosemsft This one is ready.

**Comment by danmosemsft** [comment]
@pgavlin could you review?

**Pull request merged**